CTPPS Sim and Reco: Change of paradigm: oldDD and DD4hep in parallel

ghugo83 opened the pull request:
### PR description: 

This is a **change of paradigm** of https://github.com/cms-sw/cmssw/pull/31240 , which was fully porting PPS to rely on DD4hep.

Here instead, **old DD scenarios and geometry building are restored and used by default** in CTPPS Sim and Reco.
**old DD scenarios are not affected in any way.**

DD4hep scenarios and geometry building are **added**, and can be used in CTPPS simulation and reconstruction steps instead, **in test files**, by loading the dd4hep scenarios instead (in Geometry.VeryForwardGeometry.dd4hep).

The legacy geometry building code is reshaped, so that the code for building geometry with old DD and DD4hep is in common. A fromDD4hep boolean is used when necessary.

NB: This includes few fixes (in legacy code) mentioned in https://github.com/cms-sw/cmssw/issues/31297 , at the occasion of previous PR.
NB 2 : The commits are from the last 3 days, and allow full change of paradigm.
NB 3: If any issue, time to say it now :) 
NB 4: Please keep comments related to this PR, I cannot rewrite full PPS packages legacy code unfortunately ;)

@jan-kaspar @wpcarvalho @mundim @forthommel @malbouis @fabferro @dpiparo @cvuosalo @ianna @civanch 

More details, on dedicated topics, in sections below.


### [1] Scenarios available for both oldDD and DD4hep:
The scenarios for old DD are still in Geometry/VeryForwardGeometry, and are **untouched**.
The scenarios for DD4hep are in Geometry.VeryForwardGeometry/dd4hep.

**XMLs-based scenarios**:
- geometryPPS_CMSxz_fromDD_2016_cfi
- geometryPPS_CMSxz_fromDD_2017_cfi
- geometryPPS_CMSxz_fromDD_2018_cfi
- geometryRPFromDD_2017_cfi
- geometryRPFromDD_2018_cfi

**DB scenario**:
- geometryRPFromDB_cfi


### [2] PR validation:
Checked **no regression in old DD** + Checked **no difference between oldDD and DD4hep**.

- **All workflows are untouched and rely on old DD**.

- **No diff on any geo info, when using DD4hep.** (volume names, materials, placements, shapes, sensor types, DetIds, etc).
**Fully tested on all volumes and all 6 scenarios.**
Max diff in placement is 1 um.
DD4hep is only used by test files anyway. 

- No regression in reconstruction, when using DD4hep. (tests scripts from @jan-kaspar , big thanks!):
**"Direct" simulation, using geometry from XML files**: [make_cmp_simu.pdf](https://github.com/cms-sw/cmssw/files/5124309/make_cmp_simu.pdf)
**Reconstruction of LHC data (RAW to proton), using geometry from DB**: [make_cmp_reco.pdf](https://github.com/cms-sw/cmssw/files/5124310/make_cmp_reco.pdf)
DD4hep is only used by test files anyway. 


### [3] PPS geometry building re-organization: 
PPS geometry building legacy code was heavily duplicated.
Hence, a first step was to reorganize CTPPS geometry building, to remove duplicated functionalities.
Now, generic geometry building code is used both by old DD and DD4hep.


### [4] DD4hep migration:
- Switch to DD4hep-based versions of **DDCompactView and DDFilteredView**.
- **DDTranslation → dd4hep::Direction** (ROOT::Math::XYZVector).
- **DDRotationMatrix → dd4hep::Rotation3D** (ROOT::Math::Rotation3D).
- **Fixed name casting regressions** (due to namespace not appearing with DD4hep).
- **Fixed sensor type computation regressions** (due to namespace not appearing with DD4hep). Use a SpecPar section to select the 2x2:RPixWafer volumes.
- **Fixed ID computation regressions** (due to different order of parent volumes copy numbers with DD4hep).
- **Fixed shape parameters regressions** (due to different order / units with DD4hep). Only uses the Box parameters. Struct to access data.
- **Debug and fixed units.** Since mm was the reference until now, by fear of regression, CTPPS team (with @wpcarvalho) has decided to keep mm as reference. Hence conversions as needed.

davidlange6 committed:
add missing includes

ghugo83 committed:
Remove DD4hep-specific SpecPar section from Geometry/VeryForwardData/data/CTPPS_Pixel_Module_2x2.xml. Moved into a dedicated file, only included in DD4hep scenarios.

ghugo83 committed:
change return type of PPSGeometryESProducer::applyAlignments

ghugo83 committed:
More meaningful names in PPSGeometryESProducer::applyAlignments (pD, cD, nD are hard to understand for reader). deque of observing raw pointers (ideal geo) are kept. For the aligned geo, one would need to change evth in DetGeomDesc children to be able to use smart pointers directly. For aligned geo, the children are deleted anyway, when alignedDetRoot is destructed (a DetGeomDesc object destructor calls destructors of its children DetGeomDesc).

ghugo83 committed:
Fixed exception handling of alignments.product() in PPSGeometryESProducer.cc

ghugo83 committed:
Merged dl200903 from repository davidlange6 with cms-merge-topic

ghugo83 committed:
Fixed BuildFiles + add template for old DD producers

ghugo83 committed:
Restore old DD scenarios + move DD4hep scenarios to dedicated repo.

ghugo83 committed:
Update geom print files

ghugo83 committed:
IMPORTANT: MAJOR COMMIT, restoring old DD producers, and making code consistent to have both old DD and DD4hep in parallel.

ghugo83 committed:
Restore oldDD test files + Add print geom and to DB test files for DD4hep.

ghugo83 committed:
PPSGeometryBuilder needs to have diffrent names for old DD and DD4hep.

ghugo83 committed:
Cannot add a constructor to CondFormats/GeometryObjects/interface/PDetGeomDesc.h without having a dependency mess, so just do the construction in the caller code.

ghugo83 committed:
Fix navigation

ghugo83 committed:
Fix DetId computation, to be able to work for both old DD abd DD4hep, without any code duplication. The old DD code for DetId computation is identical (apart from cleaning style).

ghugo83 committed:
Fix assignment operator (bool isDD4hep) + write computeDiamondDimensions as a private function to avoid further issues + provide DetGeomDesc comparator.

ghugo83 committed:
only complain about looking at a non-box when calling getDiamondDiemnsions

ghugo83 committed:
Renamed writer to DB + minor commnents.

ghugo83 committed:
Clearer variable naming in tree navigation + helper function in DB writer.

ghugo83 committed:
Modify ppsgeometrywriter to be able to write both DD and DD4hep in the same file

ghugo83 committed:
CTPPSGeometryESModule.cc handles both old DD and DD4hep

ghugo83 committed:
Revert "Merge pull request #31359 from cms-sw/revert-31240-CTPPS_final"

This reverts commit 2c7b8d660b81a6dd7420ec6d69c579851736de6c, reversing
changes made to 3173c7f4fe7191429f47e13c20d519d0b7617b79.

ghugo83 committed:
Merged parallel_oldDD_DD4hep from repository ghugo83 with cms-merge-topic

ghugo83 committed:
Merged parallel_oldDD_DD4hep from repository ghugo83 with cms-merge-topic

ghugo83 committed:
Simplify Geometry/VeryForwardGeometryBuilder/plugins/BuildFile.xml

ghugo83 committed:
Merged parallel_oldDD_DD4hep from repository ghugo83 with cms-merge-topic

ghugo83 committed:
scram build code-format

ghugo83 committed:
Merged parallel_oldDD_DD4hep from repository ghugo83 with cms-merge-topic

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31383/18211

- This PR adds an extra 128KB to repository

- Found files with invalid states:
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilder.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
  - Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc:
    - Modified: 490f41f5eaccf281b8d6927a41c0e39368b9a4b9, 5c1cdbb293fe6238a7a91cc99f586bf52c5fb811, 5c72fd3c02090f0918e05a5c8074bc4d623e6943
    - Deleted: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Added: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Modified: e1372d61afbae6114b231eb5e37749c551f4656e, 93db1ea00c948e53e243566a45fe3a294bccbdf5, 4e4c6be2a3ad7c382f977fabad418f521f91dbe1
    - Deleted: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/dd4hep/PPSGeometryESProducer.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: 93db1ea00c948e53e243566a45fe3a294bccbdf5
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilderDD4hep.cc:
    - Added: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
    - Modified: 6d78e24bc6afb29760c6366f43a3c72fb31084fd, 0e8e6ea50deefa66fa71fd45a4404283abb6d33b
    - Deleted: 5640c2b4b3bd7e6dffa27878a81edb074874cbb7


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DetectorDescription/DDCMS/interface/DDFilteredView.h  modified in PR(s): #31337
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2016_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2017_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2018_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometryBuilder/plugins/BuildFile.xml  modified in PR(s): #31368
  - File  Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc  modified in PR(s): #30252
  - File  Validation/CTPPS/plugins/CTPPSDirectProtonSimulation.cc  modified in PR(s): #30252

cmsbuild commented:
A new Pull Request was created by  @ghugo83  for master.

It involves the following packages:

CondFormats/GeometryObjects
CondTools/Geometry
DQM/CTPPS
DetectorDescription/DDCMS
Geometry/CMSCommonData
Geometry/VeryForwardData
Geometry/VeryForwardGeometry
Geometry/VeryForwardGeometryBuilder
RecoPPS/Local
SimPPS/RPDigiProducer
Validation/CTPPS


@perrotta, @kmaeshima, @andrius-k, @Dr15Jones, @makortel, @cvuosalo, @civanch, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @jpata, @jfernan2, @fioriNTU, @slava77, @ggovi, @pohsun, @kpedro88, @tocheng can you please review it and eventually sign? Thanks.
@vargasa, @forthommel, @jan-kaspar, @tocheng, @slomeo, @mmusich, @fabiocos, @seemasharmafnal this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 395d3dfbd6247d35c30d91749bd9abaa15cb2602
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9178/summary.html
CMSSW: CMSSW_11_2_X_2020-09-07-1100
SCRAM_ARCH: slc7_amd64_gcc820

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #30306 @kdlong: GEN-only NanoAOD configs (NanoGEN) and cmsDriver support

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9178/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9178/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9178/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2609667
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2609639
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

ghugo83 committed:
Merged parallel_oldDD_DD4hep from repository ghugo83 with cms-merge-topic

ghugo83 committed:
ESWatcher and better style in writer to DB.

ghugo83 committed:
Use ESGetToken instead of legacy way, to access data in writer to DB.

cmsbuild commented:
The code-checks are being triggered in jenkins.

ghugo83 commented:
I just added better style in the writer to DB. 

Is there anything in this PR which should be updated?
All points in https://github.com/cms-sw/cmssw/issues/31297 have been looked at and addressed.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31383/18223

- This PR adds an extra 128KB to repository

- Found files with invalid states:
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilder.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
  - Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc:
    - Modified: 490f41f5eaccf281b8d6927a41c0e39368b9a4b9, 5c1cdbb293fe6238a7a91cc99f586bf52c5fb811, 5c72fd3c02090f0918e05a5c8074bc4d623e6943
    - Deleted: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Added: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Modified: e1372d61afbae6114b231eb5e37749c551f4656e, 93db1ea00c948e53e243566a45fe3a294bccbdf5, 4e4c6be2a3ad7c382f977fabad418f521f91dbe1
    - Deleted: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/dd4hep/PPSGeometryESProducer.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: 93db1ea00c948e53e243566a45fe3a294bccbdf5
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilderDD4hep.cc:
    - Added: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
    - Modified: 6d78e24bc6afb29760c6366f43a3c72fb31084fd, 0e8e6ea50deefa66fa71fd45a4404283abb6d33b
    - Deleted: 5640c2b4b3bd7e6dffa27878a81edb074874cbb7


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DetectorDescription/DDCMS/interface/DDFilteredView.h  modified in PR(s): #31337
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2016_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2017_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2018_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometryBuilder/plugins/BuildFile.xml  modified in PR(s): #31368
  - File  Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc  modified in PR(s): #30252
  - File  Validation/CTPPS/plugins/CTPPSDirectProtonSimulation.cc  modified in PR(s): #30252

cmsbuild commented:
Pull request #31383 was updated. @perrotta, @kmaeshima, @andrius-k, @Dr15Jones, @makortel, @cvuosalo, @civanch, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @jpata, @jfernan2, @fioriNTU, @slava77, @ggovi, @pohsun, @kpedro88, @tocheng can you please check and sign again.

makortel commented on Geometry/VeryForwardGeometryBuilder/interface/DetGeomDesc.h:
```diff
@@ -30,62 +37,111 @@ class CTPPSRPAlignmentCorrectionData;
+                                             const std::vector<double>& params) const;
+  DetId computeDetID(const std::string& name, const std::vector<int>& copyNos, unsigned int copyNum) const;
+  DetId computeDetIDFromDD4hep(const std::string& name, const std::vector<int>& copyNos, unsigned int copyNum) const;
+  std::string computeSensorType(const std::string_view name);
```
Just to note that `std::string_view` provides `const`-only interface to the pointed-to string, so qualifying it as `const` is unnecessary. (technically it doesn't harm either, so I suppose the main effect is raising the question "why is it `const`")

ghugo83 replied:
Ha yes thanks, will remove the const. No reason, just historically it was a const std::string&, and then changed quickly to std::string_view because needed by sth else.

ghugo83 committed:
Fix use of std::string_view

ghugo83 commented:
Anything else?

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31383/18232

- This PR adds an extra 132KB to repository

- Found files with invalid states:
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilder.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
  - Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc:
    - Modified: 490f41f5eaccf281b8d6927a41c0e39368b9a4b9, 5c1cdbb293fe6238a7a91cc99f586bf52c5fb811, 5c72fd3c02090f0918e05a5c8074bc4d623e6943
    - Deleted: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Added: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Modified: e1372d61afbae6114b231eb5e37749c551f4656e, 93db1ea00c948e53e243566a45fe3a294bccbdf5, 4e4c6be2a3ad7c382f977fabad418f521f91dbe1
    - Deleted: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/dd4hep/PPSGeometryESProducer.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: 93db1ea00c948e53e243566a45fe3a294bccbdf5
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilderDD4hep.cc:
    - Added: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
    - Modified: 6d78e24bc6afb29760c6366f43a3c72fb31084fd, 0e8e6ea50deefa66fa71fd45a4404283abb6d33b
    - Deleted: 5640c2b4b3bd7e6dffa27878a81edb074874cbb7


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DetectorDescription/DDCMS/interface/DDFilteredView.h  modified in PR(s): #31337
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2016_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2017_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2018_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometryBuilder/plugins/BuildFile.xml  modified in PR(s): #31368
  - File  Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc  modified in PR(s): #30252
  - File  Validation/CTPPS/plugins/CTPPSDirectProtonSimulation.cc  modified in PR(s): #30252

cmsbuild commented:
Pull request #31383 was updated. @perrotta, @kmaeshima, @andrius-k, @Dr15Jones, @makortel, @cvuosalo, @civanch, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @jpata, @jfernan2, @fioriNTU, @slava77, @ggovi, @pohsun, @kpedro88, @tocheng can you please check and sign again.

ianna commented on DetectorDescription/DDCMS/interface/DDFilteredView.h:
```diff
@@ -153,41 +153,20 @@ namespace cms {
-    auto isABox(Ts&&... ts) const -> decltype(isA<dd4hep::Box>(std::forward<Ts>(ts)...)) {
-      return isA<dd4hep::Box>(std::forward<Ts>(ts)...);
-    }
+    bool isABox() const { return isA<dd4hep::Box>(); }
```
@ghugo83 - I agree with you here. There is no need to scare people with variadic templates and perfect forwarding :-)

However, it got me thinking that logically  _**"a `View` is not a `Box`"**_, but its `shape` is. These methods implement legacy API. Digging a bit more and it looks like there is not much need for them, not really. Half of them are not used and some that would be used are not implemented. The latter cases use the `dd4hep` API directly.  I'd suggest to remove these methods entirely from the `FilteredView`. 

I'd suggest:
```c++
dd4hep::isA<dd4hep::Polycone>(fv.solid());
```
instead of:
```c++
fv.isA<dd4hep::Polycone>();
```

ghugo83 replied:
Ok @ianna , will replace `fv.isA<dd4hep::Polycone>()` by `dd4hep::isA<dd4hep::Polycone>(fv.solid())` to directly use the dd4hep interface.

ghugo83 replied:
@ianna Haha yes I found it more elegant.

davidlange6 commented on Geometry/VeryForwardGeometryBuilder/interface/DetGeomDesc.h:
```diff
@@ -30,62 +37,111 @@ class CTPPSRPAlignmentCorrectionData;
-  std::vector<double> m_params;
-  DetId m_geographicalID;
+  std::string computeNameWithNoNamespace(std::string_view nameFromView) const;
+  std::vector<double> computeParameters(const cms::DDFilteredView& fv) const;
```
could this return a const & instead?

ghugo83 replied:
@davidlange6 Hmm no, this returns a local variable.

ghugo83 commented on Geometry/VeryForwardGeometryBuilder/interface/DetGeomDesc.h:
```diff
@@ -30,62 +37,111 @@ class CTPPSRPAlignmentCorrectionData;
+  // params() is left for general access to solid shape parameters, but should be used
+  // only with great care, for two reasons: 1. order of parameters may possibly change from
+  // a version to another of DD4hep; 2. length parameters unit is cm while PPS uses mm.
+  const std::vector<double>& params() const { return m_params; }  // default unit: mm from oldDD, cm from DD4hep
```
@davidlange6 The params() member function in the public interface returns const std::vector<double>& though (it returns the member data m_params).

kpedro88 commented on Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml:
```diff
@@ -348,6 +348,7 @@
     <Include ref='Geometry/VeryForwardData/data/ppstrackerMaterials/v1/ppstrackerMaterials.xml'/>
     <Include ref='Geometry/VeryForwardData/data/CTPPS_Pixel_Module/v1/CTPPS_Pixel_Module.xml'/>
     <Include ref='Geometry/VeryForwardData/data/CTPPS_Pixel_Module_2x2/v1/CTPPS_Pixel_Module_2x2.xml'/>
+    <Include ref="Geometry/VeryForwardData/data/dd4hep/SpecPar_for_CTPPS_Pixel_Module_2x2.xml"/>
```
I expect this change to fail geometry unit tests. The Run3 geometry is now generated from dictionaries, so the update should be made in: https://github.com/cms-sw/cmssw/blob/master/Configuration/Geometry/python/dict2021Geometry.py.

ghugo83 replied:
Hmm an issue I see is that this is only a DD4hep file, hence it is only added to DD4hep XML include file.

ghugo83 replied:
Basically these are specpars which only make sense to add for DD4hep.

ghugo83 replied:
If they are added to dict2021, they would also be added for old DD, which we do no want.

kpedro88 replied:
@ianna do you have any suggestions for how to handle this? We haven't encountered this case before, as far as I'm aware.

ianna replied:
I'd think it would not harm to have the file in both DD and DD4hep scenarios. The first will just ignore it.

ghugo83 replied:
Hmm in practice old DD complains if there are paths like:
`world_volume/OCMS_1/CMSE_1/RP_147_Right_Station_1/RP_147_Right_Station_Vacuum_5_1/RP_147_Right_Station_Vacuum_3_Far_1/RP_box_primary_vacuum_10103/RP_box_secondary_vacuum_0/Envelop_1/RPixWafer_1`
in SpecPar section, even if these paths are only relevant for DD4hep.
I found a trick by replacing it by 
`//RP_147_Right_Station_1/RP_147_Right_Station_Vacuum_5_1/RP_147_Right_Station_Vacuum_3_Far_1/RP_box_primary_vacuum_10103/RP_box_secondary_vacuum_0/Envelop_1/RPixWafer_1`
And in that case oldDD properly ignores it.

So I can add that to 
`Geometry/VeryForwardData/data/CTPPS_Pixel_Module_2x2/v1/CTPPS_Pixel_Module_2x2.xml`
A priori I thought it would be cleaner to not touch oldDD XMLs at all, but ok this works out.

ghugo83 replied:
Hmm actually 
`//RP_147_Right_Station_1/RP_147_Right_Station_Vacuum_5_1/RP_147_Right_Station_Vacuum_3_Far_1/RP_box_primary_vacuum_10103/RP_box_secondary_vacuum_0/Envelop_1/RPixWafer_1`
does not work either for old DD.

Hmm if we want the exact same geo dictionary, hence the exact same XMLs, hence SpecPar paths, for oldDD and DD4hep, this is gonna be tricky.
Because the volumes names in the navigation for old DD and DD4hep are slightly different.

Anyway this SpecPar section has always been a complete pain, I just went for a solution which does not use SpecPar at all.

ghugo83 replied:
I finished addressing all review points, will push the commits once I have finished re-doing all the tests.

cvuosalo commented on Geometry/VeryForwardGeometryBuilder/interface/DetGeomDesc.h:
```diff
@@ -12,10 +15,14 @@
 #include "DataFormats/DetId/interface/DetId.h"
 #include <Math/Rotation3D.h>
 
 class DDFilteredView;
```
This line can be deleted since the include has been added for `DDFilteredView`.

cvuosalo commented on Geometry/VeryForwardGeometryBuilder/src/DetGeomDesc.cc:
```diff
@@ -11,66 +11,113 @@
 
 #include "Geometry/VeryForwardGeometryBuilder/interface/DetGeomDesc.h"
 #include "Geometry/VeryForwardGeometryBuilder/interface/CTPPSDDDNames.h"
+#include "CondFormats/GeometryObjects/interface/PDetGeomDesc.h"
```
I think the following now need to be added to `Geometry/VeryForwardGeometryBuilder/BuildFile.xml`:
`CondFormats/GeometryObjects`
`DetectorDescription/Core`
`DetectorDescription/DDCMS`

cvuosalo commented on Geometry/VeryForwardGeometryBuilder/src/DetGeomDesc.cc:
```diff
@@ -11,66 +11,113 @@
+#include "CondFormats/GeometryObjects/interface/PDetGeomDesc.h"
+#include "CondFormats/PPSObjects/interface/CTPPSRPAlignmentCorrectionData.h"
 
+#include "DataFormats/Math/interface/CMSUnits.h"
```
I don't think this header is used here. It would be quite confusing if it were being used with the conflicting `GeantUnits.h`.

ghugo83 replied:
@cvuosalo Yes thanks will remove. At first was using 1. /1_mm, hence the cms units, but then was switched to geant4 units so this header will be removed.

ghugo83 committed:
clean headers

ghugo83 committed:
Updates BuildFiles: legacy SimPPS/RPDigiProducer and RecoPPS/Local were not including CondFormats/DataRecord, preventing full cleaning of Geometry/VeryForwardGeometryBuilder Buildfile.

ghugo83 committed:
Use dd4hep::isA<dd4hep::Box>(fv.solid()) instead of fv.isABox()

ghugo83 committed:
Revert "Remove DD4hep-specific SpecPar section from Geometry/VeryForwardData/data/CTPPS_Pixel_Module_2x2.xml. Moved into a dedicated file, only included in DD4hep scenarios."

This reverts commit 09e5d712da64c57e0081582e105bd04cfed6417b.

ghugo83 committed:
XMLs in old DD and DD4hep should be identical: Add DD4hep SpecPar in CTPPS_Pixel_Module_2x2.xml. The SpecPars will be ignored when running with old DD (only needed to not start the path with world_volume but with a generic //).

ghugo83 committed:
Fix in previous commit does not work: not possible to have paths which woulld be identical for old DD and DD4hep. Decided not to use SpecPar section at all, and just look at the sensor name.

ghugo83 committed:
Better DetGeomDesc printout + fix namespace access.

ghugo83 committed:
scram build code-format

cmsbuild commented:
The code-checks are being triggered in jenkins.

ghugo83 commented:
Addressed all reviews comments.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31383/18278

- This PR adds an extra 36KB to repository

- Found files with invalid states:
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilder.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
  - Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc:
    - Modified: 490f41f5eaccf281b8d6927a41c0e39368b9a4b9, 5c1cdbb293fe6238a7a91cc99f586bf52c5fb811, 5c72fd3c02090f0918e05a5c8074bc4d623e6943
    - Deleted: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Added: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Modified: e1372d61afbae6114b231eb5e37749c551f4656e, 93db1ea00c948e53e243566a45fe3a294bccbdf5, 4e4c6be2a3ad7c382f977fabad418f521f91dbe1, d6560af59cb16932750b6b3bfde916ed9df506b2, 4a484e33b22d7ea9aa8540ea3c78fe517647e9da, 74f7e3af6fc816924bb1463826d7c79afa2421c7
    - Deleted: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardData/data/dd4hep/SpecPar_for_CTPPS_Pixel_Module_2x2.xml:
    - Added: 09e5d712da64c57e0081582e105bd04cfed6417b
    - Deleted: 17d7aa62c76597106a626afaa07fdfaea12696c1
  - Geometry/VeryForwardGeometryBuilder/plugins/dd4hep/PPSGeometryESProducer.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: 93db1ea00c948e53e243566a45fe3a294bccbdf5
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilderDD4hep.cc:
    - Added: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
    - Modified: 6d78e24bc6afb29760c6366f43a3c72fb31084fd, 0e8e6ea50deefa66fa71fd45a4404283abb6d33b
    - Deleted: 5640c2b4b3bd7e6dffa27878a81edb074874cbb7


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DetectorDescription/DDCMS/interface/DDFilteredView.h  modified in PR(s): #31337
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2016_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2017_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2018_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometryBuilder/BuildFile.xml  modified in PR(s): #31368
  - File  Geometry/VeryForwardGeometryBuilder/plugins/BuildFile.xml  modified in PR(s): #31368
  - File  Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc  modified in PR(s): #30252
  - File  RecoPPS/Local/BuildFile.xml  modified in PR(s): #31368
  - File  SimPPS/RPDigiProducer/plugins/BuildFile.xml  modified in PR(s): #31368
  - File  Validation/CTPPS/plugins/CTPPSDirectProtonSimulation.cc  modified in PR(s): #30252

cmsbuild commented:
Pull request #31383 was updated. @perrotta, @kmaeshima, @andrius-k, @Dr15Jones, @makortel, @cvuosalo, @civanch, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @jpata, @jfernan2, @fioriNTU, @slava77, @ggovi, @pohsun, @tocheng can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 74f7e3af6fc816924bb1463826d7c79afa2421c7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9234/summary.html
CMSSW: CMSSW_11_2_X_2020-09-09-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9234/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2609667
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2609644
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jfernan2 commented:
+1

cvuosalo commented:
New XML directories and files should follow this naming convention:

base/filename/physical_version_key/implementation_version_key/filename.xml

Eg: Geometry/GEMGeometryBuilder/data/GEMSpecs/2019/v1/GEMSpecs.xml 

See https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideDetectorDescription#Adding_a_new_XML_file

The `Geometry/VeryForwardGeometry/data/dd4hep` directory thus needs to be changed. I am not understanding why there is a DD4hep XML directory. XML is not supposed to be DD4hep-specific.

Sorry I didn't notice this issue earlier.

cvuosalo commented:
Is the `empty.xml` file truly necessary?

cvuosalo commented:
Are any of the XML files in this PR intended to go into the Extended2021 Run 3 scenario? I am not seeing any connection to that scenario.

cvuosalo commented:
Just for my understanding, is it true that the `VeryForwardIdealGeometryRecord` has never yet been created in the Prod DB?

cvuosalo commented:
If this PR is merged, what will be the state of PPS simulation with DD4hep? Will it be enabled or will more work be needed?

christopheralanwest commented:
> Just for my understanding, is it true that the `VeryForwardIdealGeometryRecord` has never yet been created in the Prod DB?

No payloads of record type `VeryForwardIdealGeometryRecord` have been uploaded: https://cms-conddb.cern.ch/cmsDbBrowser/search/Prod/VeryForwardIdealGeometryRecord

wpcarvalho commented:
> Just for my understanding, is it true that the `VeryForwardIdealGeometryRecord` has never yet been created in the Prod DB?

Yes, it is right. It is relatively recent, introduced in the beginning of the year, as part of the effort to have PPS geometry in the database and code migrated to dd4hep.

ghugo83 commented:
>The Geometry/VeryForwardGeometry/data/dd4hep directory thus needs to be changed. I am not understanding why there is a DD4hep XML directory. 

In Geometry/VeryForwardGeometry/data/dd4hep there are the list of XMLs in include sections. Where should they be placed?

ghugo83 commented:
>Are any of the XML files in this PR intended to go into the Extended2021 Run 3 scenario? I am not seeing any connection to that scenario.
This PR is not adding any XML.
It is just adding the DD4hep scenarios.
Then there is a small volume name modification in CTPPS_Module_2x2.xml, which does not change anything (just a renaming), but allows this XML to be properly handled by the code for both oldDD and DD4hep cases.

ghugo83 commented:
>Is the empty.xml file truly necessary

Fro the DD4hep scenarios, it seems to me that the DDDetectorESProducer always ask for an XML include path, even if fromDB = true.
Why is that? Here the info comes from DB and not from XML.

ghugo83 commented:
>if this PR is merged, what will be the state of PPS simulation with DD4hep? Will it be enabled or will more work be needed?

It is fully completed.

With the previous PR, DD4hep was used everywhere for PPS instead of old DD.
Here instead, old DD is still used everywhere for PPS, and the old DD scenarios are unchanged. But there is the possibility to run evth with DD4hep instead of old DD by simply including the DD4hep scenarios instead of the old DD ones.

For example:
`process.load("Geometry.VeryForwardGeometry.dd4hep.geometryPPS_CMSxz_fromDD_2018_cfi")`
instead of:
`process.load("Geometry.VeryForwardGeometry.geometryPPS_CMSxz_fromDD_2018_cfi")`

civanch commented:
@ghugo83 , @cvuosalo , if 2021 scenario does not require modification of this PR, then better to merge and remove unwanted files in later PRs. If 2021 file needs to be moved to the right directory then better to do in this PR. 

@christopheralanwest , is this PR fine for the 2021 GT or extra things are required?

ghugo83 commented:
@civanch The PPS 2021 scenario XML files are added in another branch, this branch: 
https://github.com/ghugo83/cmssw/compare/parallel_oldDD_DD4hep...ghugo83:PPS_2021_rebase?expand=1
Diff at: https://github.com/ghugo83/cmssw/compare/parallel_oldDD_DD4hep...ghugo83:PPS_2021_rebase?expand=1 

That branch, ghugo83:PPS_2021_rebase, will only add 2021 scenarios and few fixes in the code, to be able to handle 2021 scenario smoothly.
It is based on top of this PR (DD4hep migration) to avoid massive merge conflicts, and also so that both 'old DD' and DD4hep versions of 2021 scenario can be added.

Am waiting for present PR to be merged to open (or Laurent) a PR from ghugo83:PPS_2021_rebase

cvuosalo commented:
@ghugo83  There is still the question of whether PPS XML files need to be different between old DD and DD4hep. For example, what features of `Geometry/VeryForwardGeometry/data/dd4hep/geometryRPFromDD_2018.xml` are specific to DD4hep?

I am thinking that a better name for the `Geometry/VeryForwardGeometry/data/dd4hep` directory would be `Geometry/VeryForwardGeometry/data/scenarios`, but such a change could be postponed until a later PR.

The `Geometry/CMSCommonData/data/dd4hep` directory gets its name for historical reasons. The directory name should really be "autoCreated" or something like that to indicate the files in it are generated.

> In Geometry/VeryForwardGeometry/data/dd4hep, there are just the DD4hep include sections. Hmm where should they rather be placed?
> The XMLs called are the same for old DD and DD4hep.
> 
> Basically this is the equivalent of:
> `Geometry/CMSCommonData/data/dd4hep`
> But here, with the PPS DD4hep scenarios.
> 
> To sum up:
> 
>     * All XMLs are common between old DD and DD4hep, and are at `Geometry/VeryForwardData/data`.
> 
>     * The old DD scenarios are in `Geometry/VeryForwardGeometry/python/`.
> 
>     * In this PR, DD4hep scenarios are added: at `Geometry/VeryForwardGeometry/python/dd4hep` + `Geometry/VeryForwardGeometry/data/dd4hep/ `for the includes sections.

ghugo83 commented:
>For example, what features of Geometry/VeryForwardGeometry/data/dd4hep/geometryRPFromDD_2018.xml are specific to DD4hep?

None, the XMLs used for old DD and DD4hep are now perfectly identical.
In Geometry/VeryForwardGeometry/data/dd4hep/geometryRPFromDD_2018.xml there is just the scenario definition, which for old DD is at Geometry/VeryForwardGeometry/python/geometryRPFromDD_2018_cfi.py

ghugo83 commented:
>I am thinking that a better name for the Geometry/VeryForwardGeometry/data/dd4hep directory would be Geometry/VeryForwardGeometry/data/scenarios

Ha yes true, could be like Geometry/VeryForwardGeometry/data/dd4hepScenarios

ghugo83 commented:
I think the legacy scenarios which were (and still are) at Geometry/VeryForwardGeometry/python/ are local tests, hence are not included in any dictionary.
In this PR, the scenarios which are added are just their DD4hep-equivalent local tests.

Regarding the scenarios generated by the dictionary, there is now no difference, because the DD4hep migration is transparent for the XMLs point of view.

cvuosalo commented:
I don't think `empty.xml` is needed. In `DDDetectorESProducer`, if `fromDB` is true, then `confGeomXMLFiles` is ignored. It can be set to some reasonable file name. Please remove `empty.xml` from this PR.

See below for an example of reading from DB:

https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/test/python/testGeometry2021FromDB.py?v=CMSSW_11_2_X_2020-09-08-2300#0014

ghugo83 commented:
> I don't think `empty.xml` is needed. In `DDDetectorESProducer`, if `fromDB` is true, then `confGeomXMLFiles` is ignored. It can be set to some reasonable file name. Please remove `empty.xml` from this PR.
> See below for an example of reading from DB:
> https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/test/python/testGeometry2021FromDB.py?v=CMSSW_11_2_X_2020-09-08-2300#0014

Hmm but that's the thing. Since confGeomXMLFiles is ignored when reading from DB, but must be set to sth, what should be put?

cvuosalo commented:
@ghugo83 It is not true that `Geometry/VeryForwardGeometry/python` contains only local tests. Some of the configs are used in existing workflows. That was the problem with #31240 -- it changed existing workflows by changing some of the configs in `Geometry/VeryForwardGeometry/python`.

> I think the legacy scenarios which were (and still are) at Geometry/VeryForwardGeometry/python/ are local tests, hence are not included in any dictionary.
> In this PR, the scenarios which are added are just their DD4hep-equivalent local tests.

ghugo83 commented:
Hmm there I was talking about the scenarios with XMLs - to discuss the dictionary point.

The scenarios with XMLs in Geometry/VeryForwardGeometry/python, to my understanding, are local tests and not included in the dictionary.

I think the only scenario used for outside workflows, is Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py, via 
RecoPPS/Configuration/python/recoCTPPS_cff.py.
But this one has no XML (it is the one from DB).

cvuosalo commented:
@ghugo83 
>> Since confGeomXMLFiles is ignored when reading from DB, but must be set to sth, what should be put?

How about putting `Geometry/VeryForwardGeometry/data/dd4hep/geometryRPFromDD_2018.xml`, along with a comment saying that it will be ignored?

The problem actually lies in `DDDetectorESProducer`. The correct approach is to fix it at some point so the parameter is optional. But until that happens, I don't think we should clutter up our directories with `empty.xml` files.

ghugo83 commented:
>How about putting Geometry/VeryForwardGeometry/data/dd4hep/geometryRPFromDD_2018.xml, along with a comment saying that it will be ignored?
I can if you want, but why is that better than empty.xml? Maybe the DB was generated from a 2017 xml file, I don't know.

>The problem actually lies in DDDetectorESProducer. The correct approach is to fix it at some point so the parameter is optional. But until that happens, I don't think we should clutter up our directories with empty.xml files.
Anyway, yes exactly. I was planning reporting this so that the DDDetectorESProducer gets fixed - but fixing it in this PR was out of scope I think, and adding even more snow to the snowball.

civanch commented:
@cvuosalo , Can this PR be signed and address problems in the next one?

cvuosalo commented:
I'm satisfied with this PR at this point. `empty.xml` can be removed now or in a later PR. As I said, `empty.xml` is the wrong workaround for the problem in `DDDetectorESProducer`. There is an established workaround, which is providing an existing file, though that workaround is not perfect either. At some later date, we will fix `DDDetectorESProducer` properly.

ghugo83 commented:
>At some later date, we will fix DDDetectorESProducer properly.

+1

jpata commented:
+reconstruction

- changes to reco packages are minor and understood
- reworks the previous PR https://github.com/cms-sw/cmssw/pull/31240, keeping oldDD and DD4HEP in parallel

civanch commented:
+1

civanch commented:
@ggovi,@christopheralanwest,@franzoni,@tlampen,@pohsun,@tocheng , would you , please sign this PR, which is kind of bottleneck for us. It was well discussed and now seems to be OK.

cvuosalo commented:
urgent

christopheralanwest commented:
@ghugo83 Could you please rebase this PR?

civanch commented:
@ghugo83 , rebase urgently needed.

ghugo83 committed:
Merge branch 'master' into parallel_oldDD_DD4hep

cmsbuild commented:
The code-checks are being triggered in jenkins.

ghugo83 committed:
Merged parallel_oldDD_DD4hep from repository ghugo83 with cms-merge-topic

ghugo83 commented:
I just fixed this with the online merge conflict solver, please wait till I check evth on my local.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31383/18343

- This PR adds an extra 220KB to repository

- Found files with invalid states:
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilder.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
  - Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc:
    - Modified: 490f41f5eaccf281b8d6927a41c0e39368b9a4b9, 5c1cdbb293fe6238a7a91cc99f586bf52c5fb811, 5c72fd3c02090f0918e05a5c8074bc4d623e6943
    - Deleted: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Added: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Modified: e1372d61afbae6114b231eb5e37749c551f4656e, 93db1ea00c948e53e243566a45fe3a294bccbdf5, 4e4c6be2a3ad7c382f977fabad418f521f91dbe1, d6560af59cb16932750b6b3bfde916ed9df506b2, 4a484e33b22d7ea9aa8540ea3c78fe517647e9da, 74f7e3af6fc816924bb1463826d7c79afa2421c7
    - Deleted: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardData/data/dd4hep/SpecPar_for_CTPPS_Pixel_Module_2x2.xml:
    - Added: 09e5d712da64c57e0081582e105bd04cfed6417b
    - Deleted: 17d7aa62c76597106a626afaa07fdfaea12696c1
  - Geometry/VeryForwardGeometryBuilder/plugins/dd4hep/PPSGeometryESProducer.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: 93db1ea00c948e53e243566a45fe3a294bccbdf5
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilderDD4hep.cc:
    - Added: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
    - Modified: 6d78e24bc6afb29760c6366f43a3c72fb31084fd, 0e8e6ea50deefa66fa71fd45a4404283abb6d33b
    - Deleted: 5640c2b4b3bd7e6dffa27878a81edb074874cbb7


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2016_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2017_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2018_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc  modified in PR(s): #30252
  - File  Validation/CTPPS/plugins/CTPPSDirectProtonSimulation.cc  modified in PR(s): #30252

cmsbuild commented:
Pull request #31383 was updated. @perrotta, @kmaeshima, @andrius-k, @Dr15Jones, @makortel, @cvuosalo, @civanch, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @jpata, @jfernan2, @fioriNTU, @slava77, @ggovi, @pohsun, @tocheng can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

ghugo83 committed:
Fixed merge conflict, part 2.

ghugo83 commented:
please wait am not fully done with the merge conflicts

civanch commented:
OK, please, commit.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31383/18345

- This PR adds an extra 164KB to repository

- Found files with invalid states:
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilder.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
  - Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc:
    - Modified: 490f41f5eaccf281b8d6927a41c0e39368b9a4b9, 5c1cdbb293fe6238a7a91cc99f586bf52c5fb811, 5c72fd3c02090f0918e05a5c8074bc4d623e6943
    - Deleted: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Added: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Modified: e1372d61afbae6114b231eb5e37749c551f4656e, 93db1ea00c948e53e243566a45fe3a294bccbdf5, 4e4c6be2a3ad7c382f977fabad418f521f91dbe1, d6560af59cb16932750b6b3bfde916ed9df506b2, 4a484e33b22d7ea9aa8540ea3c78fe517647e9da, 74f7e3af6fc816924bb1463826d7c79afa2421c7
    - Deleted: 581d17354502dbbb77523b6243c2557866f4b17e
  - Geometry/VeryForwardData/data/dd4hep/SpecPar_for_CTPPS_Pixel_Module_2x2.xml:
    - Added: 09e5d712da64c57e0081582e105bd04cfed6417b
    - Deleted: 17d7aa62c76597106a626afaa07fdfaea12696c1
  - Geometry/VeryForwardGeometryBuilder/plugins/dd4hep/PPSGeometryESProducer.cc:
    - Added: 5f83176c60b8d2cf4a4d3d0efb64028c7d29427c
    - Deleted: 93db1ea00c948e53e243566a45fe3a294bccbdf5
  - CondTools/Geometry/plugins/dd4hep/PPSGeometryBuilderDD4hep.cc:
    - Added: da02c2f1c5ffbf2e44c4119ab7deb0cb579f69fe
    - Modified: 6d78e24bc6afb29760c6366f43a3c72fb31084fd, 0e8e6ea50deefa66fa71fd45a4404283abb6d33b
    - Deleted: 5640c2b4b3bd7e6dffa27878a81edb074874cbb7


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2016_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2017_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometry/python/geometryIdealPPSFromDD_2018_cfi.py  modified in PR(s): #30252
  - File  Geometry/VeryForwardGeometryBuilder/plugins/CTPPSGeometryESModule.cc  modified in PR(s): #30252
  - File  Validation/CTPPS/plugins/CTPPSDirectProtonSimulation.cc  modified in PR(s): #30252

cmsbuild commented:
Pull request #31383 was updated. @perrotta, @kmaeshima, @andrius-k, @Dr15Jones, @makortel, @cvuosalo, @civanch, @tlampen, @christopheralanwest, @ianna, @mdhildreth, @cmsbuild, @jpata, @jfernan2, @fioriNTU, @slava77, @ggovi, @pohsun, @tocheng can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

ghugo83 commented:
@civanch Ok the rebase should be fine.

cmsbuild commented:
+1
Tested at: a1cb064cf1cbb8a493fd8d018ee9f03cd9ea2152
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9316/summary.html
CMSSW: CMSSW_11_2_X_2020-09-13-2300
SCRAM_ARCH: slc7_amd64_gcc820

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #31402 @mmusich: update IT CPE conditions, post dual slope ToT activation (also for other T* geometries)
 - #31452 @guitargeek: Clean BuildFiles in Fireworks

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9316/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9316/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f17aa/9316/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4635 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2620306
 - DQMHistoTests: Total failures: 54340
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2565944
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jfernan2 commented:
Are the smal (but numerous) differences seen in Phase2 expected somehow?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-09-13-2300+6f17aa/38806/dqm-histo-comparison-summary.html
Thanks

kpedro88 commented:
I suspect the Phase 2 changes occur because:
> The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic:
>
> * #31402 @mmusich: update IT CPE conditions, post dual slope ToT activation (also for other T* geometries)

jfernan2 commented:
Thanks @kpedro88  yes, the number agree

jfernan2 commented:
+1

cvuosalo commented:
+1

jpata commented:
+reconstruction

- changes to reco packages are minor and understood
- reworks the previous PR #31240, keeping oldDD and DD4HEP in parallel

(same as https://github.com/cms-sw/cmssw/pull/31383#issuecomment-691254609)

civanch commented:
+1

tlampen commented:
+alca

civanch commented:
@ggovi , can you, please, sign this PR?

ggovi commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
